src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Connection.cs

```diff
@@ -189,65 +150,45 @@ public virtual void OnSocketClosed()
             {
                 _filteredStreamAdapter.Abort();
-                _rawSocketInput.IncomingFin();
+                SocketInput.IncomingFin();
                 _readInputTask.ContinueWith((task, state) =>
                 {
                     ((Connection)state)._filterContext.Connection.Dispose();
                     ((Connection)state)._filteredStreamAdapter.Dispose();
-                    ((Connection)state)._rawSocketInput.Dispose();
+                    ((Connection)state).SocketInput.Dispose();
                 }, this);
             }
             else
             {
-                _rawSocketInput.Dispose();
+                SocketInput.Dispose();
             }
 
-            lock (_stateLock)
-            {
-                _connectionState = ConnectionState.SocketClosed;
-
-                if (_socketClosedTcs != null)
-                {
-                    // This is always waited on synchronously, so it's safe to
-                    // call on the libuv thread.
-                    _socketClosedTcs.TrySetResult(null);
-                }
-            }
+            _socketClosedTcs.TrySetResult(null);
         }
 
         private void ApplyConnectionFilter()
         {
-            lock (_stateLock)
+            if (_filterContext.Connection != _libuvStream)
             {
-                if (_connectionState == ConnectionState.CreatingFrame)
-                {
-                    _connectionState = ConnectionState.Open;
+                _filteredStreamAdapter = new FilteredStreamAdapter(ConnectionId, _filterContext.Connection, Memory, Log, ThreadPool, _bufferSizeControl);
 
-                    if (_filterContext.Connection != _libuvStream)
-                    {
-                        _filteredStreamAdapter = new FilteredStreamAdapter(ConnectionId, _filterContext.Connection, Memory, Log, ThreadPool, _bufferSizeControl);
+                _frame.SocketInput = _filteredStreamAdapter.SocketInput;
+                _frame.SocketOutput = _filteredStreamAdapter.SocketOutput;
 
-                        SocketInput = _filteredStreamAdapter.SocketInput;
-                        SocketOutput = _filteredStreamAdapter.SocketOutput;
+                _readInputTask = _filteredStreamAdapter.ReadInputAsync();
+            }
+            else
+            {
+                _frame.SocketInput = SocketInput;
+                _frame.SocketOutput = SocketOutput;
+            }
 
-                        _readInputTask = _filteredStreamAdapter.ReadInputAsync();
-                    }
-                    else
-                    {
-                        SocketInput = _rawSocketInput;
-                        SocketOutput = _rawSocketOutput;
-                    }
+            _frame.PrepareRequest = _filterContext.PrepareRequest;
 
-                    PrepareRequest = _filterContext.PrepareRequest;
+            // Reset needs to be called here so prepare request gets applied
+            _frame.Reset();
```

Comment:
I think `Frame.Start` should call `Frame.Reset` then, and we shouldn't call it in the `Frame` ctor.